packages/component/addon/index.js

```diff
@@ -153,3 +154,46 @@ export function tagName(tagName) {
     return klass;
   }
 }
+
+/**
+  Class decorator which specifies the layout for the component. This replaces
+  the `layout` property on components in the traditional Ember object model.
+
+  ```js
+  import template from '../templates/components/x-foo';
+
+  @layout(template)
+  export default class TagNameDemoComponent extends Component {}
+  ```
+
+  ```js
+  import hbs from 'htmlbars-inline-precompile';
+
+  @layout(hbs`<h1>Hello {{ name }}</h1>`)
+  export default class TagNameDemoComponent extends Component {
+    constructor() {
+      super(...arguments);
+      this.set('name', 'Tomster');
+    }
+  }
+  ```
+
+  @param {TemplateFactory} template - The compiled template to be used for the component
+*/
+export function layout(template) {
+  assert(`The @layout decorator must be provided exactly one argument, received: ${template}`, arguments.length === 1);
+  assert(`The @layout decorator must be provided a template, received: ${template}. If you want to compile strings to templates, be sure to use 'htmlbars-inline-precompile'`, typeof template !== 'string');
+  assert(`The @layout decorator must be provided a template, received: ${template}`, isTemplate(template));
+
+  return function(klass) {
+    klass.prototype.layout = template;
+    return klass;
+  }
+}
+
+/**
+ * Test to see whether a given value looks like a compiled handlebars template
+ */
+function isTemplate(val) {
```

Comment:
if you move this inline (into the assert) it will be stripped in production, I'm unsure if dead code elimination will strip it as it is now

Reply:
This began as a more complex check on the value. Do you think that checking for `type = 'object'` is good enough? We could always wrap a more complete assertion in `runInDebug`

Reply:
I was thinking you could move the function into the assert directly:

```js
assert('asdfasdfa', (() => {
  // function body here...
})());
```